kodi/Intel: Various improvements from fritsch

MilhouseVH opened the pull request:
@fritsch I've had these three commits in my Generic builds for quite a while now, originally from your OE repo. Are they all still needed? If not please let me know which (if any) should be dropped. Thanks.

piotrasd commented:
All are needed for Intel HD Gen8 like cherry trail

fritsch commented:
The kodi patch is bad for nvidia and others expecting full range.

MilhouseVH commented:
In that case I'll drop the Kodi patch.

fritsch commented:
Nope. Make it intel Auto detect, please

Am 29.07.2016 6:33 vorm. schrieb "MilhouseVH" <notifications@github.com>:

> In that case I'll drop the Kodi patch.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/LibreELEC/LibreELEC.tv/pull/573#issuecomment-236092386>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABCfHflzlPlp4qVt69iDV22c5AiBuyp-ks5qaYKIgaJpZM4JWoyY>
> .
>

MilhouseVH commented:
> Make it intel Auto detect, please

Sorry, not really sure how to accomplish that. This is the patch I dropped:

https://github.com/fritsch/OpenELEC.tv/commit/98c65e6a39a1f7ce3afc19afa8ecc25349f26031

How do we make that "Intel auto detect"?

fritsch commented:
Query the running hardware and change kodi's settings accordingly. It's no
solution already there, but I thought the best "out of the box
distribution" can accomplish that :p

2016-07-30 16:35 GMT+02:00 MilhouseVH <notifications@github.com>:

> Make it intel Auto detect, please
>
> Sorry, not really sure how to accomplish that. This is the patch I dropped:
>
> fritsch/OpenELEC.tv@98c65e6
> <https://github.com/fritsch/OpenELEC.tv/commit/98c65e6a39a1f7ce3afc19afa8ecc25349f26031>
>
> How do we make that "Intel auto detect"?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/LibreELEC/LibreELEC.tv/pull/573#issuecomment-236368457>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABCfHaYFzNtIhb8Oy9eQ1Tj74JTASj2aks5qa2ErgaJpZM4JWoyY>
> .
>



-- 
                   Key-ID:     0x1A995A9B
                   keyserver: pgp.mit.edu
==============================================================
Fingerprint: 4606 DA19 EC2E 9A0B 0157  C81B DA07 CF63 1A99 5A9B

MilhouseVH commented:
I'm not a coder (these days) so either we change the default for Intel (and possibly annoy nvidia users) or do nothing and keep the default as it is (which probably isn't great for Intel users, but hardly the end of the world). I'd vote for doing nothing - keep the current default - and merge this as-is, just the two commits.

fritsch commented:
The problem is: With the kernel patch you have, intel defaults to Video-Range 16:235.  That means signal Limited Range, but don't touch color values.

A normal TV is Limited Range by default.

Before the intel driver scaled down all input to Limited Range no matter what (!) input. Users did not notice in this combination. Now - we send Video Range by ourself without touching the color values, but kodi without the setting we discuss here, outputs full range!

That means: Everything too dark and no whites.

In short: If you set Video Range 16:235 for intel gpus, you need to set Use Limited Range, too.

MilhouseVH commented:
Forget it, I'll drop these patches from my builds. Sorry fur the noise, I thought the patches were ready to go.

fritsch commented:
Without them intel will just look bad and users need to manually fiddle with autostart.sh ... you will see tomorrow in your forum thread.

fritsch commented:
Without them the following will happen:

intel driver: scales full range to limited range
kodi: outputs full range

-> kodi will produce a full rgb image from limited content
-> kernel will just scale it down again

With above patches:
intel driver: does not care
kodi: outputs the original colors

-> kodi will just keep the colors
-> kernel will show them the way they are.

We will see if users see an effect.

MilhouseVH commented:
Like I say, not a coder so don't have the time/skill/knowledge to fix this. If the commits aren't suitable for merging upstream they have no place in the builds.

piotrasd commented:
if i good remember without this patches i had brighter picture ... like with bigger gamma or something like that, but i will check test build without this and let you know.

CvH commented:
@fritsch could you tell for the record what has to be added to the autostart.sh ?

fritsch commented:
https://github.com/LibreELEC/LibreELEC.tv/pull/573/commits/3c2175510bb494bc37eab2d5fa360f4633899d84#diff-527e71ca33779ccd70d936fc587d82e7R1 <- or a more hard coded subset ...

fritsch commented:
I want to emphasize on what I have written here:  https://github.com/LibreELEC/LibreELEC.tv/pull/573#issuecomment-236450719

Though this will happen, there is a simple solution: User just enables "Use Limited Range" and it will work as intended.

To summarize, what is missing:

It's a condition in the default settings, that if we are running on intel hardware && if we are having Video Range set && user has not altered this setting by himself set Limited Range to true.

^^ As the VideoRange patch is not in any upstream kernel this is nothing kodi in general could do. But LibreELEC that ships this patch anyways should offer such an out of the box experience for their intel user.

1) So please add the systemd patch back -> VideoRange
2) Let's think of a way to ship a custom minimal kodi patch that alters this setting on first start

MilhouseVH commented:
Thanks for the clarification. I thought that's what would happen - Intel users would just need to toggle the default setting - but all the discussion suggested otherwise.

1) This PR - in it's current form, just the 2 commits - will be added back to nightly test builds (starting with #0810 tonight)

2) I'm having trouble thinking of an easy/elegant solution. Is it possible to detect the presence of Intel hardware in `appliance.xml`?

fritsch commented:
Yes it is. See the codec settings for mpeg4 and such.

Am 10.08.2016 7:14 nachm. schrieb "MilhouseVH" <notifications@github.com>:

> Thanks for the clarification. I thought that's what would happen - Intel
> users would just need to toggle the default setting - but all the
> discussion suggested otherwise.
>
> 1) This PR - in it's current form, just the 2 commits - will be added back
> to nightly test builds (starting with #0810 tonight)
>
> 2) I'm having trouble thinking of an easy/elegant solution. Is it possible
> to detect the presence of Intel hardware in appliance.xml?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/LibreELEC/LibreELEC.tv/pull/573#issuecomment-238936716>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABCfHdBs0P1-BljJhMBE8r5aOjrDsHaKks5qegcMgaJpZM4JWoyY>
> .
>

MilhouseVH commented:
Added Intel GPU detection to Kodi. When an Intel GPU is detected, the default value for `Use limited colour range (16-235)` will be set to true - for all other GPUs it will be left as false.

The first kodi commit is based on two commits from @fritsch ([commit](https://github.com/fritsch/xbmc/commit/f6e4fef5453cfc0d5c4a28e1cc52a90865b4c34f), [commit](https://github.com/fritsch/xbmc/commit/5a3b8b7dbc276036cfd3bbdd428fed2503f2421f)) - I've simply cleaned them up, squashed them and replaced the Settings condition with the direct call to SetDefault() as it's the only way to toggle the default value (it can't be done in `appliance.xml` or `settings.xml` as there is support for conditionally changing the value of the default property).

The kodi commit works in conjunction with the systemd commit on this PR, and also the passthrough kernel [patch](https://github.com/MilhouseVH/LibreELEC.tv/blob/7cdba526c9380933894a54cf1f1a9d57e2bc041b/packages/linux/patches/4.7/linux-999-i915-implement-passthrough-colors.patch) added in Linux 4.3.0.

This is how things would normally work without all of above mentioned changes:
```
@fritsch:
20:54 so I tell you what happens
20:54 your driver is on "Automatic" which means: Limited 16:235 <- the default of intel
20:54 so the value 0 is scaled by kodi to 16
20:55 and again by the intel driver to 24
20:55 255 is scaled by kodi to 235 and again by the driver to something like 216
```
With the kernel passthrough patch we avoid the clamping in the Intel driver, however we still want to use limited range in Kodi as this is the best setting for TV displays, so limited range will now be the default for Intel GPUs. Essentially, with all these changes we will avoid clamping twice - once by Kodi and then again by the Intel driver - with Kodi dictating the range.

The downside of all this is that if the user is using LE with an Intel GPU and a computer display the new default setting will result in washed out colours as computer monitors are full range, in which case the `Use limited colour range` setting will need to be disabled by the user. On balance, there's likely to be more LibreELEC TV viewers than LibreELEC computer display viewers.

The limited range setting is currently an **Expert** level setting, but I've moved it to **Standard** with the final commit in order to make it more accessible (I will remove the final commit if requested).

I'd like to test this for a few days in my nightly builds, and if there are no issues then consider merging from Monday.

@fritsch many thanks for the time spent pulling this together!

MilhouseVH commented:
Updated to use appropriate default for VAAPI/VDPAU depending on available Intel hardware.

lrusak commented:
This looks good, thanks for all your effort here. If everything works out please just remember to update your commit dates :+1:

fritsch commented:
@lrusak if you are busy at some time - we can also think of a appliance.xml solution that detects the hardware from a systemd service or something and symlinks the proper appliance.xml per application that way no kodi hacking needs to happen.

MilhouseVH commented:
No problems reported. Ready to merge.

MilhouseVH commented:
@fritsch: What about having `appliance_hardware_<intel|nvidia|radeon|whatever>.xml` files, the specific hardware settings file being copied to `/storage/.cache/appliance_hardware.xml` by `kodi-config` according to hardware detection, and loaded as a third settings file after `appliance.xml` (ie. a [two line addition](https://github.com/xbmc/xbmc/blob/master/xbmc/settings/Settings.cpp#L841-L842) to kodi source code)? :
```
   if (CFile::Exists(SETTINGS_XML_FOLDER "appliance.xml") && !Initialize(SETTINGS_XML_FOLDER "appliance.xml"))
     CLog::Log(LOGFATAL, "Unable to load appliance-specific settings definitions");
+
+  if (CFile::Exists("/storage/.cache/appliance_hardware.xml") && !Initialize("/storage/.cache/appliance_hardware.xml"))
+    CLog::Log(LOGFATAL, "Unable to load appliance-specific hardware settings definitions");
```

This would be cleaner than having potentially multiple `appliance.xml` files, which would mostly contain duplicated appliance settings and only minor hardware specific variations. Such variations would instead only be present in the hardware specific xml file. 

Thus we could have an `appliance_hardware_intel.xml` that enables VAAPI, disabled VDPAU and enables limited range. An `appliance_hardware_nvidia.xml` could do the opposite, or alternatively we simply disable VAAPI/enable VDPAU/disable limited range in `appliance.xml` by default, and these settings would be overridden only when Intel hardware is detected.

This would be my best idea, perhaps @lrusak has other/better ideas. Although for now I'd only go this xml-based route if hardware-specific settings becomes more of a requirement in future, or Kodi changes cause us to revisit the current implementation.
